[Autoscaler] Monitor refactor for backward compatability.

AmeerHajAli opened the pull request:
In this PR I tried to optimize for backward compatibility of monitor.
Minimize direct accesses to ray library, add testing that asserts the protobuf/load metrics/etc. works

In a follow up PR I intend to add a release/e2e test for starting monitor.py in ray 1.2 versus ray master and assert autoscaling is working.


## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

rkooo567 commented on src/ray/protobuf/gcs_service.proto:
```diff
@@ -535,3 +505,34 @@ service PlacementGroupInfoGcsService {
+} // Do not change!
+
+// Service for node resource info access.
+service NodeResourceInfoGcsService { // Do not change!
```
Hmm I think it is hard to say not to change these services. (it is very much an implementation detail). Can't we just ensure backward compatibility of GetAllResourcesUsage?

AmeerHajAli replied:
I am trying to minimize risk. If the name if this service changes, monitor crashes.

AmeerHajAli replied:
If someone wants to change it, we can discuss and make sure he doesn’t do something breaking.

rkooo567 commented on src/ray/protobuf/gcs.proto:
```diff
@@ -453,3 +356,103 @@ message PubSubMessage {
+  repeated PlacementGroupTableData placement_group_data = 1; // Do not change!
+} // Do not change!
+
+message PlacementGroupTableData { // Do not change!
```
This is also hard to say it shouldn't be changed because it is literally the implementation detail of placement groups. Why don't we just parse necessary fields from these instead of marking them do not change?

AmeerHajAli replied:
If you change any field name here, or the. Name of the table, or the type of it, or add something crucial for autoscaling, it will break autoscaler/monitor.

ericl replied:
I agree with @rkooo567 here, it's not professional to add this everywhere. Let's add it one at the top of the message with a reason why.

AmeerHajAli replied:
I moved everything to autoscaler.proto

rkooo567 reviewed (commented):
In a glance, I think we shouldn't mark so many protobuf that is about implementation details "do not change". We should just have a single protobuf (API) and apply "do not change" there. We can guarantee backward compatibility in the parsing functions instead (the function that parses necessary fields from other protobuf for monitor-specific protobuf).

AmeerHajAli commented:
> In a glance, I think we shouldn't mark so many protobuf that is about implementation details "do not change". We should just have a single protobuf (API) and apply "do not change" there. We can guarantee backward compatibility in the parsing functions instead (the function that parses necessary fields from other protobuf for monitor-specific protobuf).

If I move all the rpces I marked as do not change to one autoscaler.proto and import it in gcs.proto and gcs_service.proto, will it be necessary to change all these imports in the code/how badly will it impact the existing code?

AmeerHajAli commented:
@ericl @wuisawesome , what do you think about Sang's comment and the PR in general?

ericl commented on python/ray/monitor.py:
```diff
@@ -70,6 +75,31 @@ def parse_resource_demands(resource_load_by_shape):
+    gcs_address = loop.run_until_complete(
+        _get_gcs_address(redis_address, redis_password))
+    gcs_channel = aio.insecure_channel(gcs_address)
+    return gcs_service_pb2_grpc.NodeResourceInfoGcsServiceStub(gcs_channel)
```
Why do we need asyncio stuff here, can we do it all synchronously for simplicity?

AmeerHajAli replied:
I am not sure how you would do it. It seems like the stub is by default asynchronous...

ericl replied:
Check out internal_api.py for an example of a synchronous call with the stub. The problem is you're using an AIO channel, you can instead use a normal gRPC channel.

ericl commented on python/ray/monitor.py:
```diff
@@ -117,19 +146,19 @@ def __init__(self,
+                                ) -> gcs_service_pb2.GetAllResourceUsageReply:
+        request = gcs_service_pb2.GetAllResourceUsageRequest()
+        loop = asyncio.get_event_loop()
+        return loop.run_until_complete(
```
Can we just do a synchronous call?

ericl commented on python/ray/monitor.py:
```diff
@@ -142,8 +171,10 @@ def update_load_metrics(self):
-            ip = self.raylet_id_to_ip_map.get(node_id)
+            node_id = binary_to_hex(resource_message.node_id)
+            # Make sure the node_id size never changes for backward compat.
+            assert len(resource_message.node_id) == 28
```
Please remove this assert, it prevents future changes.

AmeerHajAli replied:
Why future changes suffer here? If a future change is necessary we can modify it?

wuisawesome replied:
why is `node_id` here at all? we should just get rid of it and any references to it (including the entire raylet map)

ericl commented on python/ray/monitor.py:
```diff
@@ -364,6 +376,7 @@ def run(self):
         # Something went wrong, so push an error to all drivers.
         redis_client = ray._private.services.create_redis_client(
             args.redis_address, password=args.redis_password)
+        import ray.utils
```
?

AmeerHajAli replied:
I import it here only since it is needed when an exception happens.

AmeerHajAli replied:
Are you suggesting to remove the entire:
```
        # Something went wrong, so push an error to all drivers.
        redis_client = ray._private.services.create_redis_client(
            args.redis_address, password=args.redis_password)
        import ray.utils
        traceback_str = ray.utils.format_error_message(traceback.format_exc())
        message = ("The monitor failed with the "
                   f"following error:\n{traceback_str}")
        ray.utils.push_error_to_driver_through_redis(
            redis_client, ray_constants.MONITOR_DIED_ERROR, message)
```
it does not seem like `push_error_to_driver_through_redis` is a no op in ray.utils?

AmeerHajAli replied:
I synced with alex offline, he wanted to remove:
`traceback_str = ray.utils.format_error_message(traceback.format_exc())`

ericl commented on src/ray/protobuf/common.proto:
```diff
@@ -456,3 +443,20 @@ enum WorkerExitType {
+// backward compatibility in autoscaler.
+
+// Type of placement group strategy.
+enum PlacementStrategy { // Do not change!
```
Please remove all instances of "Do not change"

AmeerHajAli replied:
I added it to make sure the reviewer would see it, I was not sure if putting it on the top would be sufficient for the reviewer to notice.

wuisawesome reviewed (changes requested):
1. We should be deleting all the code related to the raylet map. 
2. Get rid of the "Do not change" everywhere. If you want to mark the proto's, just move them to a separate `autoscaler.proto` file and add a comment explaining why the entire file needs to be backwards compatible. 
3. Let's keep the backwards compatibility tests in a separate file. It's much easier to understand the intention. 
4. Let's get rid of all the async stuff. 

In general, this PR introduces a lot of burden on pretty much anyone touching raylet, gcs, or the autoscaler. We obviously have to be ok with some amount of burden, but we should strive to minimize it.

AmeerHajAli commented on python/ray/tests/test_multi_node_2.py:
```diff
@@ -68,16 +69,45 @@ def f():
+        assert pg_response_data.bundles[i].bundle_id.placement_group_id
+
+
+# DO NOT CHANGE THIS VERIFICATION WITHOUT NOTIFYING (Eric/Ameer/Alex).
```
this is not testing backwards_compat. this is just incrementing the existing test, to check placement groups.

wuisawesome replied:
Then why does someone need to check with us before changing a placement group feature. If anyone was to be a code owner of placement groups, it should be Sang and Luoshang

AmeerHajAli replied:
I think one of us because we are working on autoscaler.

AmeerHajAli commented:
> 1. We should be deleting all the code related to the raylet map.
> 2. Get rid of the "Do not change" everywhere. If you want to mark the proto's, just move them to a separate `autoscaler.proto` file and add a comment explaining why the entire file needs to be backwards compatible.
> 3. Let's keep the backwards compatibility tests in a separate file. It's much easier to understand the intention.
> 4. Let's get rid of all the async stuff.
> 
> In general, this PR introduces a lot of burden on pretty much anyone touching raylet, gcs, or the autoscaler. We obviously have to be ok with some amount of burden, but we should strive to minimize it.

Alex, thanks for the review.
1) 100%. I removed 95% of it but it seems as like the node_id should be removed as you pointed out.
2) Let me check this out.
3) I am not sure this is necessary since it is mostly an existing test that I added placement groups to (it tests request_resources and regular resources).
4) Sure.

AmeerHajAli commented:
I resolved all the comments.
BTW, the messages are now in autoscaler.proto. I wonder how many tests will fail because of that.
I checked and it is not really used anywhere else...

AmeerHajAli commented:
Hmm. all tests are failing.
1) I can keep the rpces where they were.
2) spend 2 more hours to fix the remaining pieces and if was not possible return to 1.

wuisawesome commented:
looks like you're close to getting the proto imports right :)

ericl commented:
Tests failing
